### PR TITLE
Add support for @(AssemblyMetadata) items that turn into assembly attributes

### LINKED ISSUE #3166: GenerateAssemblyInfo support for InternalsVisibleToAttribute and AssemblyMetadataAttribute
Hello,

In our .NET 4.7.2 applications we set the InternalsVisibleTo attribute on our libraries so that our internal unit test projects can run unit tests on internal methods without having to go through the trouble of putting public signatures of these methods into the class and wrapping everything in preprocessor directives.

We also make use of the AssemblyMetadata attribute to have our build system add vital diagnostic information to our assemblies so that there is never an issue with identifying when a project was built, from what repo commit etc.  This has made identifying and fixing bugs much easier more efficient.

To set these values we've manually set the InternalsVisibleTo attribute at design time into the AssemblyInfo.cs file.  We also have our build system inject the values into the AssemblyInfo.cs file at build time during the build (but the changes are not committed as the metadata all build version related metadata).

In .NET Standard and Core we've found that the common Assembly Attributes are now set in the project file.  Do you have plans to add these other attributes to the Microsoft.NET.GenerateAssemblyInfo.targets file so that they an be included in the project file so the new generate assembly info functionality can support adding these to the output assembly?

We have read there is a workaround where we can set the property GenerateAssemblyInfo to False and add an AssemblyInfo.cs file to the .NET Standard/Core projects but we would rather not use this workaround.  For one we do not know how long it will be supported and also we do not know if it will break the new project build workflow in some way.

Can you please provide some advice on how these attributes can be used in .NET Standard/Core project files without the workaround or an indication if you plan on adding support for them to the Microsoft.NET.GenerateAssemblyInfo.targets file?

Thank you,

Bryan

### PR DESCRIPTION
**Pull request by kzu**
Given that `AssemblyMetadataAttribute` is such a common assembly-level attribute, this adds support
for specifying it directly via simple items, such as:

```
    <AssemblyMetadata Include="Foo" Value="Bar" />
    <AssemblyMetadata Include="Bar" Value="Baz" />
```

This also avoids having to learn the `_Parameter1` and `_Parameter2` syntax in `AssemblyAttribute` elements, 
and is more similar to the way other higher-level properties like `AssemblyTitle` or `Product` are also
turned into assembly attributes.

Partially fixes #3166

The feature can be disabled by setting `$(GenerateAssemblyMetadataAttributes)` to `false`.

### CODE REVIEW

**Comment by peterhuene** [comment]
Hi @kzu.  This looks good.

Could you add some test coverage to this PR?

**Comment by kzu** [comment]
Could you point me to the place where existing tests for the other `AssemblyAttribute` items are so I can add to those? Thanks!

**Comment by peterhuene** [comment]
@kzu [here is a test](https://github.com/dotnet/sdk/blob/master/src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToControlGeneratedAssemblyInfo.cs#L36) that is setting the individual properties that cause additional `AssemblyAttribute` items to be defined.

We should probably add a test to this file to cover setting `AssemblyMetadata` items in the test project file and reflecting on the output assembly to ensure the attributes were set.

**Comment by kzu** [comment]
@peterhuene added two tests to verify both new scenarios supported. 

Thanks!

**Comment by kzu** [comment]
Added the corresponding docs PR too.

**Comment by kzu** [comment]
All green now :shipit:

**Comment by dsplaisted** [comment]
@sfoslund Would you like to review this PR?

File: src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToControlGeneratedAssemblyInfo.cs [hunk 1]
```diff
@@ -349,6 +349,65 @@ BuildCommand BuildProject(string buildNumber)
...
+                    var ns = project.Root.Name.Namespace;
+
+                    project.Root.Add(
+                        new XElement(ns + "ItemGroup",
+                            new XElement(ns + "AssemblyMetadata",
+                                new XAttribute("Include", "MetadataKey"),
+                                new XAttribute("Value", "MetadataValue"))));
+                });
+
+            new RestoreCommand(Log, testAsset.TestRoot).Execute().Should().Pass();
```
Comment (sfoslund) [inline src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToControlGeneratedAssemblyInfo.cs]: nit: I believe it's not usually necessary to explicitly restore before building anymore, is that correct @dsplaisted

File: src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToControlGeneratedAssemblyInfo.cs [hunk 2]
```diff
@@ -349,6 +349,65 @@ BuildCommand BuildProject(string buildNumber)
...
+
+            var buildCommand = new BuildCommand(Log, testAsset.TestRoot);
+            buildCommand.Execute().Should().Pass();
+
+            var assemblyPath = Path.Combine(buildCommand.GetOutputDirectory("netstandard2.0").FullName, "HelloWorld.dll");
+            var info = AssemblyInfo.Get(assemblyPath);
```
Comment (sfoslund) [inline src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToControlGeneratedAssemblyInfo.cs]: nit: variable never used

**Review by sfoslund — APPROVED** [review]
Looks good to me!

**Pull request merged**